org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/handler/ValidateActionHandler.java

```diff
@@ -47,17 +53,36 @@
 	public Object execute(ExecutionEvent event) throws ExecutionException {
 		XtextEditor xtextEditor = EditorUtils.getActiveXtextEditor(event);
 		if (xtextEditor != null) {
-			IValidationIssueProcessor issueProcessor;
 			IXtextDocument xtextDocument = xtextEditor.getDocument();
 			IResource resource = xtextEditor.getResource();
-			if(resource != null)
-				issueProcessor = new MarkerIssueProcessor(resource, xtextEditor.getInternalSourceViewer().getAnnotationModel(), markerCreator, markerTypeProvider);
-			else
-				issueProcessor = new AnnotationIssueProcessor(xtextDocument, xtextEditor.getInternalSourceViewer().getAnnotationModel(), issueResolutionProvider);
-			ValidationJob validationJob = new ValidationJob(resourceValidator, xtextDocument, issueProcessor,
-					CheckMode.ALL);
-			validationJob.schedule();
+			if(resource != null) {
+				IValidationIssueProcessor issueProcessor = new MarkerIssueProcessor(resource, xtextEditor.getInternalSourceViewer().getAnnotationModel(), markerCreator, markerTypeProvider);
+				IValidationIssueProcessor issueProcessorFastIssues = new AnnotationIssueProcessor(xtextDocument, xtextEditor.getInternalSourceViewer().getAnnotationModel(), issueResolutionProvider);
+				ValidationJob validationJob = new ValidationJob(resourceValidator, xtextDocument, new IValidationIssueProcessor() {
+					
+					@Override
+					public void processIssues(List<Issue> issues, IProgressMonitor monitor) {
+						List<Issue> fast = new ArrayList<Issue>();
```

Comment:
Theoretically we'd need to do a SubMonitor.convert(monitor) with two ticks such that we can pass the monitor consistently into the child processors. Not sure if it matters a lot, though.

Reply:
you mean 4 ticks tplit 3/1?

Reply:
I'd be ok with a 1:1 split since I assume that both list will almost always have the same number or at least almost the same number of entries (most validations are FAST).

Reply:
If we want to go all-in, that we use the combined sizes of both lists as the number of ticks.

Reply:
updated